src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,203 @@
+                _userNotificationServices.ShowWarning(failureMessage);
+                return;
+            }
+
+            // Rename the file
+            await base.RenameAsync(context, node, value); // Async process (1 of which updating Roslyn)
+
+            //Check if there are any symbols that need to be renamed
+            if (documentRenameResult.ApplicableActions.IsEmpty)
+                return;
+
+            // Ask if the user wants to rename the symbol
+            bool userWantsToRenameSymbol = await CheckUserConfirmation(oldName);
+            if (!userWantsToRenameSymbol)
+                return;
+
+            _threadingService.RunAndForget(async () =>
+            {
+                // TODO - implement PublishAsync() to sync with LanguageService
+                // https://github.com/dotnet/project-system/issues/3425)
+                // await _languageService.PublishAsync(treeVersion);
+                //
+                // Because HandleRename() is called after Roslyn, we can set this 
+                // flag when node.FilePath == HandlRename(FilePath)
+                while (_isLanguageServiceDone == false) ;
+
+                // Apply actions and notify other VS features
+                CodeAnalysis.Solution currentSolution = GetCurrentProject().Solution;
+                string renameOperationName = string.Format(CultureInfo.CurrentCulture, VSResources.Renaming_Type_from_0_to_1, oldName, value);
+                (WaitIndicatorResult result, CodeAnalysis.Solution renamedSolution) = _waitService.WaitForAsyncFunctionWithResult(
+                                title: VSResources.Renaming_Type,
+                                message: renameOperationName,
+                                allowCancel: true,
+                                token => documentRenameResult.UpdateSolutionAsync(currentSolution, token));
+
+                _roslynServices.ApplyChangesToSolution(currentSolution.Workspace, renamedSolution);
+                return;
+
+            }, _unconfiguredProject);
+
+        }
+
+        private async Task<bool> IsAutomationFunction()
+        {
+            await _threadingService.SwitchToUIThread();
+
+            _extensibility.Value.IsInAutomationFunction(out int isInAutomationFunction);
+            return isInAutomationFunction != 0;
+        }
+
+        private CodeAnalysis.Project? GetCurrentProject()
+        {
+            foreach (CodeAnalysis.Project proj in _workspace.CurrentSolution.Projects)
+            {
+                if (StringComparers.Paths.Equals(proj.FilePath, _projectVsServices.Project.FullPath))
+                {
+                    return proj;
+                }
+            }
+
+            return null;
+        }
+
+        private static CodeAnalysis.Document GetDocument(CodeAnalysis.Project? project, string? filePath) =>
+            (from d in project?.Documents where StringComparers.Paths.Equals(d.FilePath, filePath) select d).FirstOrDefault();
```

Comment:
This could be simplified to `project.Documents.FirstOrDefault(d=>StringComparers.Paths.Equals(d.FilePath, filePath)`

Reply:
We have a custom `FirstOrDefault` extension method that avoids allocating the lambda for each call too.

https://github.com/dotnet/project-system/blob/5e43903543bfe0efa835ebcc09583aa78b487d94/src/Microsoft.VisualStudio.ProjectSystem.Managed/LinqExtensions.cs#L13-L24

Reply:
done